### PR TITLE
PCHR-3657: Add filtering by absence types to leave calendar

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

This PR allows to filter by absence types in the Leave calendar by clicking on badges in the Legend.

### Filter UI in Legend component

```js
vm.absenceTypesToFilterBy // here the filter state is stored
```

Clicking on a badge toggles absence type in the filter state

```js
vm.toggleFilteringByAbsenceType = toggleFilteringByAbsenceType;
// ...
function toggleFilteringByAbsenceType (absenceTypeId) {
  if (!_.includes(vm.absenceTypesToFilterBy, absenceTypeId)) {
    vm.absenceTypesToFilterBy.push(absenceTypeId);
  } else {
    _.remove(vm.absenceTypesToFilterBy, function (_absenceTypeId_) {
      return absenceTypeId === _absenceTypeId_;
```

Changes in these filter state are watched and broadcasted.

```js
function initWatchers () {
  $rootScope.$new().$watch(function () {
    return vm.absenceTypesToFilterBy;
  }, function (newValue, oldValue) {
    if (newValue !== oldValue) {
      $rootScope.$emit('LeaveCalendar::updateFiltersByAbsenceType',
       vm.absenceTypesToFilterBy);
```

Month component listens to broadcasts and reloads when filter state has been changed

```js
$rootScope.$on('LeaveCalendar::updateFiltersByAbsenceType', function (event, _absenceTypesToFilterBy_) {
  absenceTypesToFilterBy = _absenceTypesToFilterBy_;

  showMonth(event, true);
```

In CSS we add two styles.

```css
&.chr_leave-calendar__legend__type-badge-selected {
  /* to match the mockup */
  border: solid 1px $crm-white;
  border-radius: #{$chr-leave-calendar-legend-badge-border-radius + 2px};
  box-shadow: 0 0 1px 0 $chr-leave-calendar-item-weekend-background-color;
  /* the supportive CSS - needed to support adding border without UI jumping */
  height: #{$chr-leave-calendar-legend-badge-size + 2px};
  margin-bottom: -1px;
  margin-left: -2px;
  margin-right: #{$chr-leave-calendar-legend-badge-margin - 1px};
  margin-top: #{$chr-leave-calendar-legend-badge-margin - 1px};
  padding-left: #{$chr-leave-calendar-legend-badge-padding + 1px};

&.chr_leave-calendar__legend__type-badge-not-selected {
  // not selected absence types badged should be blended
  opacity: 0.4;
```
-----
✅Manual Tests - passed
✅Jasmine Tests - augmented and passed
✅JS distributive files - included
✅Backstop tests - passed
✅CSS distributive files - included
⏹PHP Unit Tests - not needed

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-legend.component.js [hunk 1]
```diff
@@ -33,19 +37,72 @@ define([
...
+
+    /**
+     * Checks if absence type is selected for filtering
+     *
+     * @param  {String} absenceTypeId
+     * @return {Boolean}
+     */
+    function checkIfAbsenceTypeIsSelectedForFiltering (absenceTypeId) {
+      return !vm.absenceTypesToFilterBy.length ||
+        _.includes(vm.absenceTypesToFilterBy, absenceTypeId);
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-legend.component.js]: I think naming the conditions makes it more readable and helps explain why the condition is there:

```js
var isIncludedInTheAbsenceTypeFilters = _.includes(vm.absenceTypesToFilterBy, absenceTypeId);
var noAbsenceTypesFiltersHaveBeenSelected = !vm.absenceTypesToFilterBy.length;

return noAbsenceTypesFiltersHaveBeenSelected || isIncludedInTheAbsenceTypeFilters;
```

For example, it explains that this absence type is included in the filters by default if no specific absence type has been selected.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-legend.component.js [hunk 2]
```diff
@@ -33,19 +37,72 @@ define([
...
+     * @param {String} absenceTypeId
+     */
+    function toggleFilteringByAbsenceType (absenceTypeId) {
+      if (!_.includes(vm.absenceTypesToFilterBy, absenceTypeId)) {
+        vm.absenceTypesToFilterBy.push(absenceTypeId);
+      } else {
+        _.remove(vm.absenceTypesToFilterBy, function (_absenceTypeId_) {
+          return absenceTypeId === _absenceTypeId_;
+        });
+      }
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-legend.component.js]:
```js
var hasAlreadyBeenSelected = _.includes(vm.absenceTypesToFilterBy, absenceTypeId);

if (hasAlreadyBeenSelected) {
  _.remove(vm.absenceTypesToFilterBy, function (_absenceTypeId_) {
    return absenceTypeId === _absenceTypeId_;
  });
} else {
  vm.absenceTypesToFilterBy.push(absenceTypeId);
}
```

what do you think?

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/components/leave-calendar-legend.html [hunk 3]
```diff
@@ -12,10 +12,25 @@ <h2 class="panel-title chr_leave-calendar__legend__title" role="button" ng-click
       <div class="chr_leave-calendar__legend__header">Types</div>
       <div class="clearfix">
         <div ng-repeat="absenceType in legend.absenceTypes"
-          class="chr_leave-calendar__legend__type-badge"
+          ng-click="legend.toggleFilteringByAbsenceType(absenceType.id)"
+          ng-class="{
+            'chr_leave-calendar__legend__type-badge-not-selected':
+              !legend.checkIfAbsenceTypeIsSelectedForFiltering(absenceType.id),
+            'chr_leave-calendar__legend__type-badge-selected':
+              legend.checkIfAbsenceTypeIsSelectedForFiltering(absenceType.id) && legend.absenceTypesToFilterBy.length}"
```
Comment (reneolivo) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/components/leave-calendar-legend.html]: why do we need this ` && legend.absenceTypesToFilterBy.length` when the absence type has been selected? I think this was already handled by `legend.checkIfAbsenceTypeIsSelectedForFiltering`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-legend.component.js (hunk 1 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-legend.component.js]: Done

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-legend.component.js (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-calendar-legend.component.js]: yeah, better! done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/components/leave-calendar-legend.html (hunk 3 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/components/leave-calendar-legend.html]: `legend.checkIfAbsenceTypeIsSelectedForFiltering` will return TRUE if `legend.absenceTypesToFilterBy.length`, meaning, this Absence type will be **included** in the results. But in this particular case we need to apply class if ONLY it was selected manually.

**Pull request merged**